src/emqx_os_mon.erl

```diff
@@ -189,19 +223,93 @@ ensure_check_timer(State = #{cpu_check_interval := Interval}) ->
         _ -> State#{timer := emqx_misc:start_timer(timer:seconds(Interval), check)}
     end.
 
-%% At startup, memsup starts first and checks for memory alarms,
-%% but emqx_alarm_handler is not yet used instead of alarm_handler,
-%% so alarm_handler is used directly for notification (normally emqx_alarm_handler should be used).
-%%The internal memsup will no longer trigger events that have been alerted,
-%% and there is no exported function to remove the alerted flag,
-%% so it can only be checked again at startup.
-ensure_system_memory_alarm(HW) ->
-    case erlang:whereis(memsup) of
-        undefined -> ok;
-        _Pid ->
-            {Total, Allocated, _Worst} = memsup:get_memory_data(),
-            case Total =/= 0 andalso Allocated/Total * 100 > HW of
-                true -> emqx_alarm:activate(high_system_memory_usage, #{high_watermark => HW});
-                false -> ok
-            end
+ensure_mem_check_timer(#{mem_check_timer := Ref} = State) when is_reference(Ref) ->
+    %% timer already started
+    State;
+ensure_mem_check_timer(State) ->
+    Interval = resolve_mem_check_interval(State),
+    case is_sysmem_check_supported() of
+        true ->
+            State#{mem_check_timer => emqx_misc:start_timer(timer:seconds(Interval), check_mem),
+                   mem_check_interval => Interval
+                  };
+        false ->
+            State
+    end.
+
+resolve_mem_check_interval(#{mem_check_interval := Seconds}) when is_integer(Seconds) ->
+    Seconds;
+resolve_mem_check_interval(_) ->
+    %% this only happens when hot-upgrade from older version (< 4.3.14, or < 4.4.4)
```

Comment:
This is also in call stack of `init/1` I think.

Reply:
yes it is, but there the interval seconds is set before calling,  i.e. matches the first clause of this function.